### DISCUSSION TITLE
In OpenSSL 3.1.4, where is the definition of MD5_Update and MD5_Final?

### DISCUSSION THREAD

**Discussion by Syllinia** [openssl/openssl#25489] [Q&A] [answered]
The below 2 places didn't define them either.

md5_prov.c
/* ossl_md5_functions */
IMPLEMENT_digest_functions(md5, MD5_CTX,
                           MD5_CBLOCK, MD5_DIGEST_LENGTH, 0,
                           MD5_Init, MD5_Update, MD5_Final)

legacy_md5.c
IMPLEMENT_LEGACY_EVP_MD_METH(md5, MD5)

**Comment by levitte** [comment, marked as answer]
It's a bit complicated.  Many of the hash implementations made using [`include/crypto/md32_common.h`](https://github.com/openssl/openssl/blob/master/include/crypto/md32_common.h), and then the individual implementations define the bits that are unique to them, as well as what `HASH_UPDATE`, `HASH_FINAL` etc should really be called.

If you look in [`crypto/md5/md5_local.h`](https://github.com/openssl/openssl/blob/master/crypto/md5/md5_local.h), you will see how all of this is defined for MD5 specifically.

**Reply by Syllinia** [reply]
Awesome. Using macro substitution to implement template function in C++.
